# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def add_two(x):
    return x + 2
print(add_two(3))

#Lambda function:
lambda x: x + 2

(lambda x: x + 2)(3)


5

#### Saving to a Variable

In [5]:
f = lambda x,y : x + y + 2
(lambda x,y : x + y + 2)(2,3)
f(2,3)

7

#### Multiple Inputs

In [8]:
(lambda x,y,z : x * y * z)(3,5,8)
three_multiply = lambda x,y,z : x * y * z
three_multiply = (3,5,8)

120

#### Passing a Lambda into a Function

In [9]:
def multiply (f,num):
    return f(num)
multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [11]:
def return_fun():
    def multiply(num):
        return num * num
    return multiply
# return fun() represents multiply
return_fun()(5)


def return_lamb(b,c):
    return lambda x,a : x + a+ b + c

return_lamb(2,4)(5,7)

25

#### If Statements within Lambdas

In [13]:
#Lambda x : True if (condition) else false
f = lambda num : num * 2 if num > 10 else num + 2

f(10)
f(12)

24

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [15]:
def cube():
    def cube_cude(num):
        return num ** 3
    return cube_cude

cube()(2)


f = lambda num : num ** 3
f(2)

8

In [16]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [18]:
sorted([2,7,34534,67567,45,2233,78,8])

[2, 7, 8, 45, 78, 2233, 34534, 67567]

In [24]:
j = {
    "name":"Jesus",
    "age":6
}
t = {
    "name":"Tate",
    "age":18
}
m = {
    "name":"McCall",
    "age":23
}
a = {
    "name":"Anu",
    "age":97
}

people = [m,j,a,t]
print(people)
# lambda function must return the value you want to sort on 
print(sorted(people, key = lambda person_dict: person_dict["age"]))

print(sorted(people, key = lambda person_dict: person_dict["age"], reverse = True))


[{'name': 'McCall', 'age': 23}, {'name': 'Jesus', 'age': 6}, {'name': 'Anu', 'age': 97}, {'name': 'Tate', 'age': 18}]
[{'name': 'Jesus', 'age': 6}, {'name': 'Tate', 'age': 18}, {'name': 'McCall', 'age': 23}, {'name': 'Anu', 'age': 97}]
[{'name': 'Anu', 'age': 97}, {'name': 'McCall', 'age': 23}, {'name': 'Tate', 'age': 18}, {'name': 'Jesus', 'age': 6}]


In [ ]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [25]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [29]:
def squared(num, num2):
    if num < 10 and num2 <10:
        return num**2, num2**2
    else:
        return num, num2
    
numbers = [4,10,3,2,6]
more_nums = [4,11,20,3,15]
squared_nums_map = map(squared, numbers, more_nums)
list(squared_nums_map)


[(16, 16), (10, 11), (3, 20), (4, 9), (6, 15)]

#### Using Lambda's with Map

In [27]:
numbers = [4,10,3,2,6]
squared_nums = list(map(lambda x: x ** 2 if x < 10 else x,numbers))
squared_nums


[16, 10, 9, 4, 36]

In [33]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))


squared_num2 = list(map(lambda x1, y1: (x1**2, y1**2) if x1 <10 and y1 < 10 else (x1,y1),numbers,more_nums))
squared_num2

[(16, 16), (10, 11), (3, 20), (4, 9), (6, 15)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [52]:
nums = [4,2,6,8]
dub_minus_one = list(map(lambda x: (x * 2) - 1, nums))
dub_minus_one

[7, 3, 11, 15]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [53]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [58]:
# filter the list for first names starting with T
names = ["Jesus", "Tim", "Christina", "Eduardo", "David", "Tate"]

def t_names(name):
    if name[0].lower() == "t":
        return True
    else:
        return False
new_names_list = filter(t_names, names)
list(new_names_list)

['Tim', 'Tate']

#### Using Lambda's with Filter()

In [61]:
new_names_lamb = list(filter(lambda name: True if name[0].lower() == 't' else False, names))
new_names_lamb 

['Tim', 'Tate']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [60]:
from statistics import mean
mean([1,2,3])

2

In [75]:
nums = [2,7,4.2,1.6,9,4.4,4.9]

new_mean = list(filter(lambda name: True if nums >= mean(nums) else False,nums))
new_mean 



TypeError: '>=' not supported between instances of 'list' and 'float'

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [66]:
from functools import reduce
# reduce(function,iterable)
help(reduce)


Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [68]:
list_1 = [2,4,6,8,10]

def subtract_nums (num1, num2):
    return num1 - num2
result = reduce(subtract_nums, list_1)
result

-26

#### Using Lambda's with Reduce()

In [71]:
l_3 =  [1,10,2,30]
result_lamb = reduce(lambda x,y: x - y, l_3)
result_lamb

-41

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [73]:
l_5 = [1,2,3,4]
mul_lamb = reduce(lambda x,y: x * y, l_5)
mul_lamb

24

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [74]:
def recursive(i):
    print("hi")
    if i == 4:
        return
    i += 1
    recursive(i)

recursive(0)

hi
hi
hi
hi
hi


#### Implementing a Base Case

In [84]:
def add_nums(num):
    #set base case
    if num <= 1:
        print("add_nums(1) = 1")
        return num
    else:
        print(f"add_num({num}) = {num} + add_nums({num - 1})")
        x = num -1
        return num + add_nums(x)
add_nums(5)

add_num(5) = 5 + add_nums(4)
add_num(4) = 4 + add_nums(3)
add_num(3) = 3 + add_nums(2)
add_num(2) = 2 + add_nums(1)
add_nums(1) = 1


15

#### Writing a Factorial Function

In [88]:
# 5! = 5*4*3*2*1

def factorial(num):
    if num <= 1:
        print("factorial(1) = 1")
        return 1
    else:
        print(f"factorial({num}) + factorial({num-1})")
        return num * factorial(num-1)

factorial(50)

factorial(50) + factorial(49)
factorial(49) + factorial(48)
factorial(48) + factorial(47)
factorial(47) + factorial(46)
factorial(46) + factorial(45)
factorial(45) + factorial(44)
factorial(44) + factorial(43)
factorial(43) + factorial(42)
factorial(42) + factorial(41)
factorial(41) + factorial(40)
factorial(40) + factorial(39)
factorial(39) + factorial(38)
factorial(38) + factorial(37)
factorial(37) + factorial(36)
factorial(36) + factorial(35)
factorial(35) + factorial(34)
factorial(34) + factorial(33)
factorial(33) + factorial(32)
factorial(32) + factorial(31)
factorial(31) + factorial(30)
factorial(30) + factorial(29)
factorial(29) + factorial(28)
factorial(28) + factorial(27)
factorial(27) + factorial(26)
factorial(26) + factorial(25)
factorial(25) + factorial(24)
factorial(24) + factorial(23)
factorial(23) + factorial(22)
factorial(22) + factorial(21)
factorial(21) + factorial(20)
factorial(20) + factorial(19)
factorial(19) + factorial(18)
factorial(18) + factorial(17)
factorial(

30414093201713378043612608166064768844377641568960512000000000000

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [90]:
n = 5
def subtract(num):
    if num <= 1:
        print("1")
        return 1
    else:
        print(f"{num} - subtract({num}-1)")
        return num - subtract(num-1)
subtract(5)

5 - subtract(5-1)
4 - subtract(4-1)
3 - subtract(3-1)
2 - subtract(2-1)
1


3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [91]:
list(range(5))

[0, 1, 2, 3, 4]

In [96]:
def my_range(n):
    for i in range(5):
        yield i
        
list(my_range(5))

[0, 1, 2, 3, 4]

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [102]:
def my_range(start, stop, step=1):
    while start < stop:
        yield start
        start += step
list(my_range(1,5,1))

for i in my_range(0,5):
    print(i)

0
1
2
3
4


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [104]:
def range_squared(start, stop, step=1):
    while start > stop:
        yield start ** 2
        start -= step
for num in range_squared(0,10,0):
    print(num)
    

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego",""," ","","Boston","New York"]

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`